make codegen'd device guards not cuda-specific. Allow them to be used in external codegen

bdhirsh opened the pull request:
Right now, we generate device guards + device checks in the codegen, but only for CUDA/HIP. This PR generalizes that to work with external backends that want to use device guards (Intel). It:
- exposes a `device_guard` flag in the yaml that an external backend can set
- Uses that flag to determine whether or not generate device guards + checks. For in-tree backends, we set that flag using the old method: Only set it for CUDA/HIP.

I also sanity tested this by adding the `device_guard` line into xla's yaml file, and confirming that device guards + device checks were generated in `RegisterXLA.cpp`. I also confirmed that cuda codegen output is unchanged.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69686
* __->__ #68531
* #68530
* #68529

Differential Revision: [D32498570](https://our.internmc.facebook.com/intern/diff/D32498570)

bdhirsh committed:
make codegen'd device guards not cuda-specific. Allow them to be used in external codegen

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32498570).

guoyejun commented:
verified it works, thanks.

albanD commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -339,20 +339,25 @@ def generate_defn(cpp_sig: CppSignature) -> str:
+                        device_guard = """
   const DeviceGuard device_guard(device_or_default(device));"""
+
+                        # CUDA requires special handling
```
Other backend should be able to optin to this lazy init as well? But that can be done later.

bdhirsh replied:
yep that's right - it seems like in the long term we'd want some sort of unified API for backends to perform lazy init. Intel doesn't currently need one, so we can push that off for later.

albanD commented on tools/codegen/model.py:
```diff
@@ -646,6 +646,10 @@ class BackendIndex:
     # Mainly important for structured kernels, this determines which variant in the operator group is used to implement the others.
     # All in-tree ops use out kernels, while XLA uses functional kernels.
     use_out_as_primary: bool
+    # Whether the backend requires a device guard, and device checks.
```
extra comma?

albanD commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -339,20 +339,25 @@ def generate_defn(cpp_sig: CppSignature) -> str:
 
                 device_check = '  // No device check\n'
-                if is_cuda_dispatch_key(self.backend_index.dispatch_key):
+                # Backends that require device guards presumably also require device checks.
```
nit remove the presumably? It is documented as also providing the device check.

albanD reviewed (commented):
Looks good!
Did you verify that the core codegened code didn't change?

bdhirsh committed:
Update on "make codegen'd device guards not cuda-specific. Allow them to be used in external codegen"

Right now, we generate device guards + device checks in the codegen, but only for CUDA/HIP. This PR generalizes that to work with external backends that want to use device guards (Intel). It:
- exposes a `device_guard` flag in the yaml that an external backend can set
- Uses that flag to determine whether or not generate device guards + checks. For in-tree backends, we set that flag using the old method: Only set it for CUDA/HIP.

I also sanity tested this by adding the `device_guard` line into xla's yaml file, and confirming that device guards + device checks were generated in `RegisterXLA.cpp`. I also confirmed that cuda codegen output is unchanged.


Differential Revision: [D32498570](https://our.internmc.facebook.com/intern/diff/D32498570)

[ghstack-poisoned]

albanD commented:
Some lint issues need to be fixed though.

bdhirsh committed:
Update on "make codegen'd device guards not cuda-specific. Allow them to be used in external codegen"

Right now, we generate device guards + device checks in the codegen, but only for CUDA/HIP. This PR generalizes that to work with external backends that want to use device guards (Intel). It:
- exposes a `device_guard` flag in the yaml that an external backend can set
- Uses that flag to determine whether or not generate device guards + checks. For in-tree backends, we set that flag using the old method: Only set it for CUDA/HIP.

I also sanity tested this by adding the `device_guard` line into xla's yaml file, and confirming that device guards + device checks were generated in `RegisterXLA.cpp`. I also confirmed that cuda codegen output is unchanged.


Differential Revision: [D32498570](https://our.internmc.facebook.com/intern/diff/D32498570)

[ghstack-poisoned]

bdhirsh committed:
Update on "make codegen'd device guards not cuda-specific. Allow them to be used in external codegen"

Right now, we generate device guards + device checks in the codegen, but only for CUDA/HIP. This PR generalizes that to work with external backends that want to use device guards (Intel). It:
- exposes a `device_guard` flag in the yaml that an external backend can set
- Uses that flag to determine whether or not generate device guards + checks. For in-tree backends, we set that flag using the old method: Only set it for CUDA/HIP.

I also sanity tested this by adding the `device_guard` line into xla's yaml file, and confirming that device guards + device checks were generated in `RegisterXLA.cpp`. I also confirmed that cuda codegen output is unchanged.


Differential Revision: [D32498570](https://our.internmc.facebook.com/intern/diff/D32498570)

[ghstack-poisoned]

bdhirsh committed:
Update on "make codegen'd device guards not cuda-specific. Allow them to be used in external codegen"

Right now, we generate device guards + device checks in the codegen, but only for CUDA/HIP. This PR generalizes that to work with external backends that want to use device guards (Intel). It:
- exposes a `device_guard` flag in the yaml that an external backend can set
- Uses that flag to determine whether or not generate device guards + checks. For in-tree backends, we set that flag using the old method: Only set it for CUDA/HIP.

I also sanity tested this by adding the `device_guard` line into xla's yaml file, and confirming that device guards + device checks were generated in `RegisterXLA.cpp`. I also confirmed that cuda codegen output is unchanged.


Differential Revision: [D32498570](https://our.internmc.facebook.com/intern/diff/D32498570)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32498570).

bdhirsh committed:
Update on "make codegen'd device guards not cuda-specific. Allow them to be used in external codegen"

Right now, we generate device guards + device checks in the codegen, but only for CUDA/HIP. This PR generalizes that to work with external backends that want to use device guards (Intel). It:
- exposes a `device_guard` flag in the yaml that an external backend can set
- Uses that flag to determine whether or not generate device guards + checks. For in-tree backends, we set that flag using the old method: Only set it for CUDA/HIP.

I also sanity tested this by adding the `device_guard` line into xla's yaml file, and confirming that device guards + device checks were generated in `RegisterXLA.cpp`. I also confirmed that cuda codegen output is unchanged.


Differential Revision: [D32498570](https://our.internmc.facebook.com/intern/diff/D32498570)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32498570).
